### PR TITLE
Adding automatic lazy bundling, issue #87

### LINKED ISSUE #87: Support automatic lazy widget bundling
**Enhancement**

When using `@dojo/core/load` to lazy load a module, we want automatically add the module to its own bundle

Given,

```typescript
registry.define('my-widget', () => {
    return load(require, './MyWidget')
        .then(([ MyWidget ]) => MyWidget.default);
});
```

We want `cli-build` to add `MyWidget` to it's own bundle using something like,

```typescript
require(`bundle-loader?lazy&name=${widgetName}!./Widget`);
```

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

This adds automatic bundling for the lazy widget loading pattern.  The pattern looks like this,

```typescript
import load from '@dojo/core/load';

registry.define('some-widget', () => {
    return load(require, './widgets/SomeWidget').then(([someWidget]) => someWidget.default);
});
```

Detecting this pattern will automatically create a new output file that will get loaded dynamically.

*About the pattern*

The pattern being looked for is a method call where the first parameter is a `require` identifier, the second parameter is a `string`, and the entire method call is wrapped inside of a `.define` method call.

You can disable bundling with the `--disableLazyWidgetDetection` option.

Resolves #87

### CODE REVIEW

File: src/plugins/CoreLoadPlugin.ts [hunk 1]
```diff
@@ -54,11 +61,91 @@ function stripPath(basePath: string, path: string): string {
...
+	ignoredModules?: string[];
+	basePath?: string;
+}
+
 /**
  * A webpack plugin that forces webpack to ignore `require` passed as a value, and replaces `@dojo/core/load` with a
  * custom function that maps string module IDs to webpack's numerical module IDs.
  */
 export default class DojoLoadPlugin {
+	detectLazyLoads: boolean;
```
Comment (agubler) [inline src/plugins/CoreLoadPlugin.ts]: Are these properties `private`?

File: src/plugins/CoreLoadPlugin.ts [hunk 2]
```diff
@@ -54,11 +61,91 @@ function stripPath(basePath: string, path: string): string {
...
+	lazyChunkNames: DojoLoadChunkNames;
+	ignoredModules: Set<string>;
+
+	constructor(options: DojoLoadPluginOptions = {}) {
+		const { detectLazyLoads, chunkNames, ignoredModules, basePath = '' } = options;
```
Comment (agubler) [inline src/plugins/CoreLoadPlugin.ts]: If you default the options here you won't need the `||` when initialising the properties.

File: src/plugins/CoreLoadPlugin.ts [hunk 3]
```diff
@@ -54,11 +61,91 @@ function stripPath(basePath: string, path: string): string {
...
+
+		this.detectLazyLoads = detectLazyLoads || false;
+		this.lazyChunkNames = chunkNames || {};
+		this.ignoredModules = new Set<string>();
```
Comment (agubler) [inline src/plugins/CoreLoadPlugin.ts]: Can initialise this inline in the property declaration (if you wanted).

File: src/plugins/CoreLoadPlugin.ts [hunk 4]
```diff
@@ -54,11 +61,91 @@ function stripPath(basePath: string, path: string): string {
...
+
+		if (ignoredModules) {
+			ignoredModules.forEach(moduleName => {
+				const absolutePath = path.resolve(basePath, moduleName);
+
+				this.ignoredModules.add(absolutePath);
+			});
+
+			console.log(this.ignoredModules);
```
Comment (agubler) [inline src/plugins/CoreLoadPlugin.ts]: Did we mean to log the ignored modules here?

File: src/plugins/util/parser.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ * Determine whether the specified node is shadowing any of the specified variable names.
+ *
+ * @param item
+ * The node to test
+ *
+ * @param importNames
+ * A list of variable names.
+ *
+ * @return
+ * `true` if the
```
Comment (agubler) [inline src/plugins/util/parser.ts]: if the..... it's a mystery :)

File: src/plugins/util/parser.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+ * Determine whether the provided node is a function declaration.
+ *
+ * @param item
+ * The item to test.
+ *
+ * @return
+ * `true` if the node represents either function declaration or a function expression.
+ */
+export function isFunctionDefinition(item: any): item is BaseFunction {
+	const { type } = item;
```
Comment (agubler) [inline src/plugins/util/parser.ts]: Can `item` be undefined? you have guarded against it in `isArrayExpression`

File: src/plugins/util/parser.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		'BlockStatement',
+		'ForInStatement',
+		'ForOfStatement',
+		'ForStatement',
+		'FunctionDeclaration',
+		'FunctionExpression',
+		'LetStatement',
+		'Program',
+		'TryStatement'
+
```
Comment (agubler) [inline src/plugins/util/parser.ts]: line break?

File: src/plugins/util/parser.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	].forEach((type: string) => {
+		nextItemMap[ type ] = getters.body;
+	});
+
+	return function (item: any): any {
```
Comment (agubler) [inline src/plugins/util/parser.ts]: Can `item` be undefined?

File: src/plugins/CoreLoadPlugin.ts [hunk 9]
```diff
@@ -72,17 +159,141 @@ export default class DojoLoadPlugin {
...
+												const edep = new RequireEnsureItemDependency(second.value, second.range);
+												edep.loc = dep.loc;
+												dep.addDependency(edep);
+											});
+
+											/*
+											 By default, the require.ensure is not going to execute when we want. We wrap it in a function block
+											 to control the execution.
+											 */
+											(<any> old).addDependency(new ConstDependency('function() { return (', fnExpression.range[0]));
```
Comment (agubler) [inline src/plugins/CoreLoadPlugin.ts]: As you have to cast all the usages would it better to `const old: any = parser.state.current;` (if that works) and then just cast back to the current type when it gets reassigned? `parser.state.current = <CurrentType> old;`?

File: src/plugins/CoreLoadPlugin.ts [hunk 10]
```diff
@@ -1,10 +1,17 @@
 import * as path from 'path';
+import { getBasePath, resolveMid } from './util/main';
+import { CallExpression, Program } from 'estree';
+import { getNextItem } from './util/parser';
+import Set from '@dojo/shim/Set';
```
Comment (agubler) [inline src/plugins/CoreLoadPlugin.ts]: As this runs in Node do we need the shim?

File: src/plugins/CoreLoadPlugin.ts (hunk 4 above)
Comment (rorticus) [inline src/plugins/CoreLoadPlugin.ts]: Nope... oops 😄

File: src/plugins/CoreLoadPlugin.ts (hunk 9 above)
Comment (rorticus) [inline src/plugins/CoreLoadPlugin.ts]: Removed the `<any>` casts everywhere for `old`

File: src/plugins/util/parser.ts (hunk 7 above)
Comment (rorticus) [inline src/plugins/util/parser.ts]: Yup!

File: src/plugins/util/parser.ts (hunk 8 above)
Comment (rorticus) [inline src/plugins/util/parser.ts]: Item shouldn't be undefined unless something is wrong with the AST tree, but I added a check regardless.

File: src/plugins/util/parser.ts (hunk 6 above)
Comment (rorticus) [inline src/plugins/util/parser.ts]: Item shouldn't be undefined unless something is wrong with the AST tree, but I added a check regardless.

File: src/plugins/util/parser.ts (hunk 5 above)
Comment (rorticus) [inline src/plugins/util/parser.ts]: Haha.

**Comment by rorticus** [comment]
The tests dont cover the webpack config at all, so I manually tested the population of `ignoredModules` from the `bundles` config.

Widget loading,

```typescript
registry.define('todo-footer', () => {
	return load(<any> require, './widgets/TodoFooter').then(([ todoFooter ]) => todoFooter.default);
});

registry.define('view-chooser', () => {
	return load(<any> require, './widgets/ViewChooser').then(([ viewChooser ]) => viewChooser.default);
});

registry.define('filters', () => {
	return load(<any> require, './widgets/TodoFilter').then(([ todoFilter ]) => todoFilter.default);
});
```

`.dojorc`:

```json
{
	"build-webpack": {
		"bundles": {
			"widgets": [
				"src/widgets/ViewChooser",
				"src/widgets/TodoFilter"
			]
		}
	}
}
```

`dojo build` then generates the following files,

* `main.js`
* `widgets.js` (based on bundle information in `.dojorc`)
* `TodoFooter.js` (based on automatic detection)

**Comment by rorticus** [comment]
Added a check for `@dojo/core/load` before processing the rest of the AST. Should speed things up a bit.

File: src/plugins/CoreLoadPlugin.ts (hunk 10 above)
Comment (mwistrand) [inline src/plugins/CoreLoadPlugin.ts]: I had to do the same thing with `Map` in `I18nPlugin` due to type collisions in what was either an update to TS or to `@dojo/shim` (I can't remember exactly which).

**Pull request merged**